R/navs-legacy.R

```diff
@@ -105,36 +105,236 @@ navset_hidden <- function(..., id = NULL, selected = NULL,
-                     bg = NULL, inverse = "auto",
-                     collapsible = TRUE, fluid = TRUE) {
+navset_bar <- function(
+  ...,
+  title = NULL,
+  id = NULL,
+  selected = NULL,
+  sidebar = NULL,
+  fillable = TRUE,
+  gap = NULL,
+  padding = NULL,
+  header = NULL,
+  footer = NULL,
+  fluid = TRUE,
+  navbar_options = NULL,
+  position = deprecated(),
+  bg = deprecated(),
+  inverse = deprecated(),
+  collapsible = deprecated()
+) {
   padding <- validateCssPadding(padding)
   gap <- validateCssUnit(gap)
 
-  navs_bar_(
-    ..., title = title, id = id, selected = selected,
-    sidebar = sidebar, fillable = fillable,
-    gap = gap, padding = padding,
+  .navbar_options <- navbar_options_resolve_deprecated(
+    options_user = navbar_options,
     position = position,
-    header = header, footer = footer,
-    bg = bg, inverse = inverse,
-    collapsible = collapsible, fluid = fluid,
+    bg = bg,
+    inverse = inverse,
+    collapsible = collapsible
+  )
+
+  navs_bar_(
+    ...,
+    title = title,
+    id = id,
+    selected = selected,
+    sidebar = sidebar,
+    fillable = fillable,
+    gap = gap,
+    padding = padding,
+    header = header,
+    footer = footer,
+    fluid = fluid,
+    position = .navbar_options$position,
+    bg = .navbar_options$bg,
+    inverse = .navbar_options$inverse,
+    collapsible = .navbar_options$collapsible,
+    underline = .navbar_options$underline %||% FALSE,
     # theme is only used to determine whether legacy style markup should be used
     # (and, at least at the moment, we don't need legacy markup for this exported function)
     theme = bs_theme()
   )
 }
 
+#' Create a set of navbar options
+#' 
+#' A `navbar_options()` object captures options specific to the appearance and
+#' behavior of the navbar, independent from the content displayed on the page.
+#' This helper should be used to create the list of options expected by
+#' `navbar_options` in [page_navbar()] and [navset_bar()].
+#' 
+#' ## Changelog
+#' 
+#' This function was introduced in \pkg{bslib} v0.9.0, replacing the `position`,
+#' `bg`, `inverse`, `collapsible` and `underline` arguments of [page_navbar()]
+#' and [navset_bar()]. Those arguments are deprecated with a warning and will be
+#' removed in a future version of \pkg{bslib}.
+#' 
+#' @examples
+#' navbar_options(position = "static-top", bg = "#2e9f7d", underline = FALSE)
+#' 
+#' @inheritParams shiny::navbarPage
+#' @param bg a CSS color to use for the navbar's background color.
+#' @param inverse Either `TRUE` for a light text color or `FALSE` for a dark
+#'   text color. If `"auto"` (the default), the best contrast to `bg` is chosen.
+#' @param underline Whether or not to add underline styling to page links when
+#'   active or focused.
+#' 
+#' @returns Returns a list of navbar options.
+#' 
+#' @export
+navbar_options <- function(
+  position = NULL,
+  bg = NULL,
+  inverse = NULL,
+  collapsible = NULL,
+  underline = NULL
+) {
```

Comment:
It feels slightly strange to introduce a new function in a file named `navs-legacy.R` -- maybe `R/page.R` or a new file would be better?

Also, is the reason why we don't provide default so that we can throw an informative warning when both are supplied? If yes, that informative warning doesn't seem worth the added complexity and harder-to-discover defaults.

Reply:
I think `navs-legacy.R` is misnamed? This function is defined just under `navset_bar()`, which IMHO makes it a pretty good place to put the definition.



> Also, do we not provide defaults here so that we can throw an informative warning when both are supplied?

Not just an informative warning, but also so that the current implementation can continue to work. If we move the defaults into `navbar_options()` we won't be able to tell the difference between user using an argument in `navbar_options = navbar_options()` and the deprecated direct argument version -- because the defaults in `navbar_options()` will always take precedence over the old argument.

So it's mostly about keeping existing code working during a transition period where the primary warning would mostly be the deprecation warning.

That said, we could just go all the way and hard-deprecate the old arguments. Or we could do even more complicated meta programming. The current approach intends to be a middle ground.

Reply:
Ah right, ok, thanks for pointing that out. One more half-baked thought -- could we surface defaults in the function signature but set change them to `NULL` when `missing()`?

Reply:
FWIW, we could also mark `missing()` arguments with a special class, like:

```r
arg <- if (missing(arg)) default(arg) else arg

default <- function(x) {
  class(x) <- c("bslib-default-value", class(x))
  x
}

"%|D|%" <- function(x, y) {
  if (is.default(x)) y else x
}
```

which is essentially how `ggplot2::wavier()` works, which addresses a similar problem

Reply:
Thanks, I had considered this but thought it'd be adding complexity. That said, the hidden defaults (from a user perspective) bugs me and I think the trade-off could be worth it. I'll try this out now.